File: 백준/Silver/2583. 영역 구하기/영역 구하기.py
'''
BFS를 통한 풀이
boolean 맵에 겹친 후, False인 것들 그룹 체크
'''
import sys
from collections import deque

# 가능한 범위 체크 함수
def is_valid(x,y):
    return 0<=x<M and 0<=y<N

delta_x = [0,0,-1,1]
delta_y = [1,-1,0,0]

N,M,K = map(int, sys.stdin.readline().split())
graph = []
for _ in range(N):
    graph.append([False]*M)
for _ in range(K):
    x1,y1,x2,y2 = map(int, sys.stdin.readline().split())
    for i in range(y1,y2):
        for j in range(x1,x2):
            graph[i][j] = True # 계속 겹치기
visited = []
for _ in range(N):
    visited.append([False]*(M))

res = []
for i in range(N):
    for j in range(M):
        if not graph[i][j] and not visited[i][j]:
            deq = deque()
            deq.append([j,i])
            visited[i][j] = True
            cnt = 1
            while deq:
                cur = deq.popleft()
                for k in range(4):
                    dx = cur[0] + delta_x[k]
                    dy = cur[1] + delta_y[k]
                    if is_valid(dx,dy) and not graph[dy][dx] and not visited[dy][dx]:
                        deq.append([dx,dy])
                        visited[dy][dx] = True
                        cnt += 1
            res.append(cnt)

res.sort()
print(len(res))
print(*res)

File: 백준/Silver/2583. 영역 구하기/영역 구하기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Deque;
import java.util.List;
import java.util.StringTokenizer;

public class Main {
	static BufferedReader input = new BufferedReader(new InputStreamReader(System.in));
	static StringTokenizer token;
	static StringBuilder sb = new StringBuilder();
	
	static int m,n,k;
	static boolean[][] graph, visited;
	static List<Integer> resArr;
	
	static int[] deltaX = {0,0,1,-1};
	static int[] deltaY = {1,-1,0,0};


	public static void main(String[] args) throws IOException {
		token = new StringTokenizer(input.readLine());
		m = Integer.parseInt(token.nextToken());
		n = Integer.parseInt(token.nextToken());
		k = Integer.parseInt(token.nextToken());
		
		graph = new boolean[m][n];
		visited = new boolean[m][n];
		for (int t=0; t<k; t++) {
			token = new StringTokenizer(input.readLine());
			int x1 = Integer.parseInt(token.nextToken());
			int y1 = Integer.parseInt(token.nextToken());
			int x2 = Integer.parseInt(token.nextToken());
			int y2 = Integer.parseInt(token.nextToken());
			
			for (int i=y1; i<y2; i++) {
				for (int j=x1; j<x2; j++) {
					graph[i][j] = true;
				}
			}
		}
		resArr = new ArrayList<>();
		for (int i=0; i<m; i++) {
			for (int j=0; j<n; j++) {
				if (!visited[i][j] && !graph[i][j]) {
					int cnt = 0;
					Deque<int[]> deq = new ArrayDeque();
					deq.addLast(new int[] {j,i});
					cnt++;
					visited[i][j] = true;
					
					while(!deq.isEmpty()) {
						int[] cur = deq.pollFirst();
						int x = cur[0];
						int y = cur[1];
						
						for (int d=0; d<4; d++) {
							int dx = x+deltaX[d];
							int dy = y+deltaY[d];
							
							if (isValid(dx, dy) && !graph[dy][dx] && !visited[dy][dx]) {
								deq.addLast(new int[] {dx,dy});
								cnt++;
								visited[dy][dx] = true;
							}
						}
						
					}
					resArr.add(cnt);
				}
			}
		}
		Collections.sort(resArr);
		sb.append(resArr.size()).append("\n");
		for (int a : resArr) {
			sb.append(a).append(" ");
		}
		System.out.println(sb);
		
	}
	public static boolean isValid(int x, int y) {
		return x>=0 && y>=0 && x<n && y<m;
	}
}